### PR TITLE
Type annotate some hookspecs & impls

### PR DESCRIPTION
**Pull request by bluetech**
Extracted from #6717.

Annotate some "easy" arguments of hooks that repeat in a lot of internal plugins.

Not all of the arguments are annotated fully for now.

### CODE REVIEW

File: src/_pytest/cacheprovider.py [hunk 1]
```diff
@@ -372,16 +376,18 @@ def pytest_addoption(parser):
...
 
         return wrap_session(config, cacheshow)
+    return None
 
 
 @pytest.hookimpl(tryfirst=True)
-def pytest_configure(config):
-    config.cache = Cache.for_config(config)
+def pytest_configure(config: Config) -> None:
+    # Type ignored: pending mechanism to store typed objects scoped to config.
```
Comment (bluetech) [inline src/_pytest/cacheprovider.py]: This repeats in several plugins. I have a simple plan for how to do it, but for now I added type-ignores.

File: src/_pytest/logging.py [hunk 2]
```diff
@@ -510,13 +512,13 @@ def __init__(self, config):
         )
 
         log_file = get_option_ini(config, "log_file")
-        if log_file:
+        if not log_file:
```
Comment (bluetech) [inline src/_pytest/logging.py]: Swapped the order so mypy can infer the type without an annotation.

File: src/_pytest/hookspec.py [hunk 3]
```diff
@@ -185,15 +198,15 @@ def pytest_collection_modifyitems(session, config, items):
...
+def pytest_collection_finish(session: "Session"):
     """ called after collection has been performed and modified.
 
     :param _pytest.main.Session session: the pytest session object
     """
 
 
 @hookspec(firstresult=True)
-def pytest_ignore_collect(path, config):
+def pytest_ignore_collect(path, config: "Config"):
```
Comment (blueyed) [inline src/_pytest/hookspec.py]: `path: py.path.local`

Comment (bluetech) [reply src/_pytest/hookspec.py]: I prefer leaving annotating `path` out of this particular PR.

**Pull request closed without merging**